appsec.go

```diff
@@ -0,0 +1,259 @@
+	AppSecConfig *AppSecConfig `yaml:"appsec_config"`
+	Client       *http.Client  `yaml:"-"`
+}
+
+type AppSecResponse struct {
+	Response   *http.Response
+	Action     string `json:"action"`
+	HTTPStatus int    `json:"http_status"`
+}
+
+// Config() fills the struct with configuration values from a file. It is not
+// aware of .yaml.local files so it is recommended to use ConfigReader() instead.
+func (w *AppSec) Config(configPath string) error {
+	reader, err := os.Open(configPath)
+	if err != nil {
+		return fmt.Errorf("unable to read config file '%s': %w", configPath, err)
+	}
+
+	return w.ConfigReader(reader)
+}
+
+func (w *AppSec) ConfigReader(configReader io.Reader) error {
+	content, err := io.ReadAll(configReader)
+	if err != nil {
+		return fmt.Errorf("unable to read configuration: %w", err)
+	}
+
+	err = yaml.Unmarshal(content, w)
+	if err != nil {
+		return fmt.Errorf("unable to unmarshal config file: %w", err)
+	}
+
+	return nil
+}
+
+func (w *AppSec) Init() error {
+	var err error
+
+	if w.AppSecConfig.Url == "" {
+		return fmt.Errorf("config does not contain AppSec url")
+	}
+
+	if w.AppSecConfig.ParsedUrl, err = url.Parse(w.AppSecConfig.Url); err != nil {
+		return fmt.Errorf("unable to parse AppSec url: %w", err)
+	}
+
+	if w.AppSecConfig.InsecureSkipVerify == nil {
+		w.AppSecConfig.InsecureSkipVerify = new(bool)
+		*w.AppSecConfig.InsecureSkipVerify = false
+	}
+
+	w.AppSecConfig.Timeout.SetDefaults()
+
+	caCertPool, err := getCertPool(w.AppSecConfig.CAPath, log.StandardLogger())
+	if err != nil {
+		return err
+	}
+
+	dialContext := (&net.Dialer{
+		Timeout: time.Duration(*w.AppSecConfig.Timeout.ConnectTimeout) * time.Second,
+	}).DialContext
+
+	if w.AppSecConfig.ParsedUrl.Scheme == "unix" || strings.HasPrefix(w.AppSecConfig.ParsedUrl.String(), "/") {
+		dialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return net.Dial("unix", strings.TrimSuffix(w.AppSecConfig.ParsedUrl.Path, "/"))
+		}
+		w.AppSecConfig.ParsedUrl.Host = "unix"
+		w.AppSecConfig.ParsedUrl.Scheme = "http"
+	}
+
+	w.Client = &http.Client{
+		Transport: &http.Transport{
+			DialContext: dialContext,
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: *w.AppSecConfig.InsecureSkipVerify,
+				RootCAs:            caCertPool,
+			},
+			TLSHandshakeTimeout:   time.Duration(*w.AppSecConfig.Timeout.TLSHandshakeTimeout) * time.Second,
+			ResponseHeaderTimeout: time.Duration(*w.AppSecConfig.Timeout.ResponseHeaderTimeout) * time.Second,
+		},
+	}
+
+	return nil
+}
+
+// ParseClientReq parses the client request and returns a new request that is ready to be forwarded to the AppSec.
+// You can override the IP address with the ipOverride parameter.
+// This function should not be used directly, use Forward() instead.
+func (w *AppSec) ParseClientReq(ctx context.Context, clientReq *http.Request, ipOverride string) (*http.Request, error) {
+	var req *http.Request
+
+	if clientReq.Body != nil && clientReq.ContentLength > 0 {
```

Comment:
We need some kind of override for the body: in some situations, we *do not* want to read the entire body in memory (for example, a file upload of a few GB).
To keep things simple in the library, it could just be a true/false flag, and we let the user set it (main drawback being this means different bouncers are very likely to have different configuration/support).

Cleaner solution would be to provide additional configuration with for example body size, URL prefix, source IP and so on

(this is an issue we already have in the nginx appsec integration)